File: benchmarks/17_fibonacci_big/fibonacci.c
/*
 * Fibonacci(1,000,000) - Big Integer Benchmark
 * Tests arbitrary precision arithmetic and iterative computation
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <gmp.h>

int main() {
    const int N = 1000000;
    mpz_t a, b, temp;
    
    printf("Computing Fibonacci(%d)...\n", N);
    
    mpz_init_set_ui(a, 0);
    mpz_init_set_ui(b, 1);
    mpz_init(temp);
    
    for (int i = 0; i < N; i++) {
        mpz_set(temp, a);
        mpz_add(a, a, b);
        mpz_set(b, temp);
    }
    
    // Get number of digits for verification
    size_t digits = mpz_sizeinbase(a, 10);
    printf("Fibonacci(%d) has %zu digits\n", N, digits);
    
    // Get first 10 digits for verification
    char *str = mpz_get_str(NULL, 10, a);
    printf("First 10 digits: %.10s...\n", str);
    printf("Last 10 digits: %.10s...\n", str + digits - 10);
    free(str);
    
    mpz_clear(a);
    mpz_clear(b);
    mpz_clear(temp);
    
    return 0;
}

File: benchmarks/17_fibonacci_big/fibonacci.go
/*
 * Fibonacci(1,000,000) - Big Integer Benchmark
 * Tests arbitrary precision arithmetic and iterative computation
 */
package main

import (
	"fmt"
	"math/big"
	"time"
)

func main() {
	const N = 1000000
	
	fmt.Printf("Computing Fibonacci(%d)...\n", N)
	
	start := time.Now()
	
	a := big.NewInt(0)
	b := big.NewInt(1)
	temp := new(big.Int)
	
	for i := 0; i < N; i++ {
		temp.Set(a)
		a.Add(a, b)
		b.Set(temp)
	}
	
	elapsed := time.Since(start)
	
	// Get number of digits for verification
	digits := len(a.String())
	fmt.Printf("Fibonacci(%d) has %d digits\n", N, digits)
	
	// Get first and last digits for verification
	str := a.String()
	fmt.Printf("First 10 digits: %s...\n", str[:10])
	fmt.Printf("Last 10 digits: %s...\n", str[len(str)-10:])
	
	fmt.Printf("Time: %v\n", elapsed)
}
